### DISCUSSION TITLE
how much should be done for the user

### DISCUSSION THREAD

**Discussion by MatthewCaseres** [acturtle/cashflower#212] [General]
Some aspects of this project I feel that too much is imposed on the user, in terms of explaining how things should be set up.

In some places there are these wrappers for dataframes and I wonder if we really need the wrapper. Another thing in the docs is it says

> Assumptions are stored as a dictionary. Each item in the dictionary is a data frame.

Why do things have to be like that? What if my assumption is a dataclass and it contains numpy arrays? Can I not use this software?

I see a lot of opportunity in this recursive systems analyzers, but I feel that the focus in this project is specific to a certain project setup and am hoping the software can be generally beneficial even outside an actuarial context.

**Comment by zchmielewska** [comment]
Hi @MatthewCaseres, thanks for your comment!

In general, the idea of this project is to create a framework which is an open-source alternative to expensive commercial products (like Prophet or Risk Agility FM). I see a framework as a certain structure to solve a problem, so that's true that the focus is on the structure and how to edit its components.

You gave a very good example, where I have unnecessarily imposed the given structure. You are right, we can use a dictionary to store any kind of data. Thanks for this remark and I'll update the docs. We can set an assumption dictionary like that:

```
assumption = {
    "interest_rate": 0.01,
    "mortality_rate": pd.DataFrame({
        "age": [10, 20, 30],
        "rate": [0.001, 0.002, 0.003]
    }),
    "inflation": np.array([0.12, 0.11, 0.10])
}
```

In another part of your comment, you say:

> In some places there are these wrappers for dataframes 

What some places do you mean? For example, model point set and runplan?

```
runplan = Runplan(data=pd.DataFrame({"version": [1]}))
main = ModelPointSet(data=pd.DataFrame({"id": [1]}))
```

Yes, currently it needs to be a `pandas` data frame because later on, the model performs data transformation and uses functions and methods from `pandas` package (e.g.`.loc` and `concat`).

Or did you mean any other example? What data type would you like to use instead (e.g. `numpy` matrix)?
If the data frame is not imposed, then the software would need to check what kind of data type it is and behave accordingly.

My background is actuarial so I focus on the actuarial context from the user perspective but I also see the application of the package in a more general financial context (valuation of assets and banking products such as mortgages).

What other opportunities do you see in the recursive systems analyzers?

Thanks again for your comment, especially about the `assumption` dictionary.

**Reply by MatthewCaseres** [reply]
## Being general

Like you could do a good visualization of fibonnaci with memoization for example with this recursive systems analyzer. knapsack DP too? Really any top-down memoization DP. There is an aspect of what you are doing which is very general.

I want the decorator to be good and maybe not restricting me to single argument with the letter t. And I want a good dependency viewer and if code navigation can work within vscode I think it is really nice.

## Code is specific

When I see instructions in the docs like
>runplan.get("my-column")

Why define a dataframe wrapper and then use this special `get` method to get the column? I could get the column without putting it in a custom object. Overall I would rather have the dataframe than a wrapper over it.

I just don't want your library to do too many things because then it might not be flexible and it imposes the cognitive overhead of me having to understand all of the things you are doing for me.

**Reply by zchmielewska** [reply]
**Being general**

I see your point and I see the advantages of being general. But in this case, I have a very particular target group in mind and my aim is to create a fit-for-purpose tool for this group. I don't know if that's a good approach or not. Maybe a general tool would be more useful but it's difficult to create something without knowing the needs of the end-user (who they are? what they do? and why they do it?).

The cash flow models predict the future and the `t` variable reflects future periods. That's why the decorator requires the function with the `t` parameter. It's exactly the same in the commercial software and I think that's a good approach.

I'll definitely look into VSCode functionalities.

**Code is specific**

Cash flow models are usually run for multiple versions. For example, the base version, then the version with stressed mortality and another version with stressed lapses. The runplan contains configuration for each of these versions. The user can start the given version while running:

```python run.py 2```

This command would start the second version. I created a class `Runplan` because then the user doesn't have to worry about connecting the version they are running with the correct row in the runplan. The `.get()` method does it for them. If runplan was a plain data frame, then the user would have to think and program the connection between version and configuration. 

What would be your preferred solution to running the model for multiple versions of assumptions?

I agree that the lower the cognitive overhead, the better. Software for cash flow models is usually complex so I try to keep it as simple as possible. If you think something could be easier, I'd be happy to hear!

**Reply by MatthewCaseres** [reply]
There are some benchmarks of open source life insurance software. We have many implementations of the lifelib basic term model, here is an example of a decorator implementation: https://github.com/actuarialopensource/benchmarks/blob/main/Python/basicterm_scratch.py

This implementation imposes few restrictions and is similar to the `@cache` decorator except that it is possible to invalidate all of the caches simultaneously.

**Why I will want to use your software instead of a simpler decorator?**

And apologies if I come here to complain about things, it is exciting that you are working on this project. I very much want to see a project of this type succeed, which is why I try to bring up reasons I'm currently hesitant to use it.

**Reply by zchmielewska** [reply]
The most important thing about `basicterm_scratch.py` is that I did two changes to the assumptions and it does not calculate anymore:

- `max_proj_len` - it's set to 20 years but that's a low number for life insurance; usually, the expected prediction horizon is at least 60 years and maximally 120 years; here I set it to 100 years (`max_proj_len = 12 * 100 + 1`),
- `mp` - I increased the number of model points from 10_000 (which is rather a small portfolio) to 100_000.

I have an 8GB memory computer and the script throws an error `IndexError`. The script tries to calculate results for all model points at the same time which is too memory-expensive with bigger datasets.

Also, writing functions that return arrays is a bigger cognitive load than just focusing on relations between variables (especially for non-tech actuaries).

By all means, don't apologise, I am very happy that you have written and I really appreciate that you challenge the project. You have also suggested really nice things!

**Reply by MatthewCaseres** [reply]
interesting, I will have to make sure to explore more modelpoints over longer periods of time. I haven't done this sort of modeling professionally. Do you have any references for more information on things like prediction horizon or how many modelpoints?

**Reply by zchmielewska** [reply]
Where I work current commercially, the company uses either 60 or 120 years as a projection horizon. I think 120 years is a maximal sensible projection horizon because it's usually assumed that 120 years is a maximal age (after that mortality rate = 100%). 

When it comes to model points, the company has now closed books (they don't sell any more, so it used to be more...). The model with the highest number of policies has ~315k policies. The smallest one has ~5k policies. However, I think that a tool for cash flow models should work for any number of model points. It will be slower for a bigger number of policies, but in general, it should not be limited by the input.

**Reply by MatthewCaseres** [reply]
Many thanks, these numbers are very helpful.

> I think that a tool for cash flow models should work for any number of model points

We can't handle more policies than will fit in the memory of the computer for example, like if the modelpoints dataframe is too large. Someone is working on improving memory efficiency in our benchmarks, but they are using pure Julia and it is not in a way that actuaries will be familiar with.

I will need to implement your software in the benchmarks repository and do some stress testing under different conditions to really see what is happening.

**Reply by acturtle** [reply]
I'm looking a little bit more into the script that you have sent me. I have one more remark - I wanted to add more variables to the output because I compare our results. 

I added:

```
@cash
def result_cf():
    t_len = range(max_proj_len)
    data = {
        "duration": [np.sum(duration(t)) for t in t_len],
        "pols_death": [np.sum(pols_death(t)) for t in t_len],
        "pols_if": [np.sum(pols_if(t)) for t in t_len],
        "lapse_rate": [np.sum(lapse_rate(t)) for t in t_len],
        "pols_lapse": [np.sum(pols_lapse(t)) for t in t_len],
        "pols_maturity": [np.sum(pols_maturity(t)) for t in t_len],


        "premiums": [np.sum(premiums(t)) for t in t_len],
        "claims": [np.sum(claims(t)) for t in t_len],
        "expenses": [np.sum(expenses(t)) for t in t_len],
        "commissions": [np.sum(commissions(t)) for t in t_len],
        "net_cf": [np.sum(net_cf(t)) for t in t_len]
    }
    return pd.DataFrame(data, index=t_len)
```

and I receive:

```
RecursionError: maximum recursion depth exceeded in comparison
```

The problem appeared when I added `pols_maturity`. 

Maybe the problem is that these variables form a cycle:
- `pols_death`
- `pols_if`
- `pols_lapse`
- `pols_maturity`

and because of that, Python is calling many functions inside of each other and finally the RecursionError appears.
I am not sure why caching doesn't do the trick here... But in cashflower, I try to find these cycles, isolate them and calculate together.

**Reply by zchmielewska** [reply]
Sorry, I was logged in to a repository's account but it was still me - Zuzanna :)

**Reply by MatthewCaseres** [reply]
Nice catch. The issue is calling `pols_maturity(0)` which calls `pols_if(-1)` which goes down and down forever. 

Here are some random thoughts, you may be aware of these things but I just like to write - 

You might encounter overhead associated with maintaining this graph for cycle detection. If you are calculating these policies individually then there will be graph computations for each policy, but an array based approach sort of gets around this at the cost of memory consumption.

These recursive life insurance models experience exponential time complexity without memoization, basically a dynamic programming problem similar to fibonnaci sequence calculation.

In the fibonacci problem we can do it in the following ways
- top down (recursion with memoization), O(N) space O(N) time
- bottom up (allocate results array for all the numbers up to N), O(N) space O(N) time
- bottom up (allocate only the two current number), O(1) space O(N) time

It is almost always easiest to use top down with memoization, and almost always hardest to write some memory efficient bottom up implementation. And so actuarial calculation engines in my view are like a top down dynamic programming thing.

For the purpose of the benchmarks someone is working on bottom-up memory efficient implementations in Julia, which I am supposing will be very scalable and performant, but a little harder to read.

**Reply by zchmielewska** [reply]
Hi @MatthewCaseres, sorry for the long reply, I was off for some time. 

I looked into the 3  ways to solve fibonacci that you mentioned.

Are these the examples of implementation that you have in mind?

0. basic recursive fibonacci
```
def fibonacci(n):
    if n == 0 or n == 1:
        return n

    return fibonacci(n-1) + fibonacci(n-2)
```

1. top down (recursion with memoization), O(N) space O(N) time
```
def fibonacci(n, cache):
    if n not in cache:
        if n == 0 or n == 1:
            cache[n] = n
        else:
            cache[n] = fibonacci(n-1, cache) + fibonacci(n-2, cache)

    return cache[n]
```

2. bottom up (allocate results array for all the numbers up to N), O(N) space O(N) time
```
def fibonacci(n):
    lst = [None for _ in range(n+1)]

    for i in range(n+1):
        if i == 0 or i == 1:
            lst[i] = i
        else:
            lst[i] = lst[i-1] + lst[i-2]

    return lst[n]
```

3. bottom up (allocate only the two current number), O(1) space O(N) time
```
def fibonacci(n):
    count = 1
    a = 0
    b = 1

    while count < n:
        c = a + b
        a, b = b, c
        count += 1

    return c
```

Please let me know if that's what you had in mind because I have some question.

**Reply by MatthewCaseres** [reply]
Yeah that is exactly what I was thinking about. There is also some recent discussion of vectorized vs. seriatim models on the modelx blog - https://modelx.io/blog/2023/08/19/enhanced-speed-for-exported-lifelib-models/

**Reply by zchmielewska** [reply]
Thanks, I'll take a look into this blog post!

I have few thoughts on the 3 ways you mentioned.

The **way 3** (bottom up with two concurrent numbers) - it seems like a trick because of the nature of the fibonacci's problem. I don't think it can be generalized. [?!]

The **way 1** (top down with memoization) - it raises a `RecursionError` for bigger `n`.

```
print(fibonacci(1440, {}))
```

results in:

```
RecursionError: maximum recursion depth exceeded in comparison
```

So I think the **way 2** (bottom up with array) makes most sense (or some kind of its variation). 

---

BUT it's worth taking into consideration that some actuarial variables are recursive backward and some forward.

For example:

- backward - calendar month (it uses `cal_month(t-1)`):

```
def cal_month(t):
    if t == 0:
        return runplan.get("valuation_month")
    if cal_month(t-1) == 12:
        return 1
    else:
        return cal_month(t-1) + 1
```

- forward - most of the present value calculations (it uses `actuarial_present_value(t+1)`)

```
def actuarial_present_value(t):
    if t == settings["T_MAX_CALCULATION"]:
        return expected_payment(t)
    return expected_payment(t) + actuarial_present_value(t+1) * 1/(1+INTEREST_RATE)
```

So depending on the variable "bottom up" can become "top down".

**Reply by zchmielewska** [reply]
@MatthewCaseres, just in case you are interested, I have coded the Basic Term also in cashflower.

It is available in the repository [https://github.com/acturtle/modelling](https://github.com/acturtle/modelling) in the folder `lifelib_cashflower/basic_term`.

Generally the structure is:

- `input.py` - model points, assumptions and parameters,
- `model.py` - model's logic,
- `settings.py` - some settings regarding running and output structure,
- `run.py` - script to run the model.

If you source the `run.py` script, the results will be saved in the output folder as a csv file.

**Reply by MatthewCaseres** [reply]
Awesome, really helpful so I can better see how the cashflower system works because I'm familiar with the lifelib model. Many thanks
